### ISSUE TITLE
Funtion lbl to label patches in a matrix.

### ISSUE THREAD

**Issue by alaindanet** [fdschneider/spatial_warnings#11]
- Change the structure of the fonction in the standard way as it was proposed by [Alex](https://github.com/fdschneider/spatial_warnings/issues/6).

- Add the documentation

**Comment by ssumithra** [comment]
Done

**Comment by fdschneider** [comment]
The dependency of this function needs to be defined properly. Instead of using `require()` the package must be called as a dependency via the DESCRIPTION and the function must be imported in the  NAMESPACE.  

I am still having trouble getting the `EBImage::bwlabel()` function to work. The package is hosted on Bioconductor and is not straightforward to install, at least in my Ubuntu set-up. Calling that package as a dependency is problematic. Maybe there is an alternative, or we can get our own labelling function written in C?

**Comment by ssumithra** [comment]
Ah I see. Didn't know about the calling format, sorry.

What are we going to do about EBImage? Ages ago Sabiha wrote the bwlabel
function in C but it was ridiculously slow and we thus abandoned it. Could
try to dig that up, but I think it may be better to write afresh given that
it was a failed attempt.

**Comment by fdschneider** [comment]
We already have an [R function](https://github.com/fdschneider/caspr/blob/master/R/patches.R) that works well for landscape objects, but it is too slow. Maybe it is easy for you to translate it into C for binary matrices? You can use the Rcpp framework, as [Alex did for the predprey update function](https://github.com/fdschneider/caspr/blob/master/src/predprey_core.cpp).

**Comment by fdschneider** [comment]
I just pushed a version that runs in R. It depends on `caspr::mapping()` which is not ideal. 
If we can replace `label()` with a version written in C, that would be perfect.

**Comment by alexgenin** [comment]
It does not look so hard to implement in c++, I can have a look at it, but probably not before the end of this week unfortunately.

**Comment by fdschneider** [comment]
That would be awesome. As I said for now we can run the R-function and substitute it later  with a faster one. It should have  
- a switch from 4-cell to 8-cell neighborhood or even better handle an alternative interaction matrix (which could be asymmetrical).
- switch for periodic *or* absorbing boundaries

**Comment by alexgenin** [comment]
OK, I added several functions (some of them would be useful for caspr too) in commit f75fc5c6114791bfd6e2195d88d270e9b17afe08 : 
 - get_neighbors_coords/get_neighbors_values : returns the x/y coordinates of a cell in a matrix or the values of neighbords
 - label : label patches in a binary matrix

They all use a "neighborhood mask" and have a switch for absorbing or periodic boundaries (wrap argument). More details are available in the doc of these functions. 

A quick example (from label's doc): 
```
data(B)
image(label(B))
# With 8-way neighborhood mask
nbmask8 <- matrix(c(1,1,1,
                    1,0,1,
                    1,1,1), ncol=3, byrow=TRUE)
image(label(B, nbmask8, wrap = FALSE))
```

I also updated indicator_cumpsd, psv (patch size vector) and psd (patch size distribution) to reflect the changes.

I tried my best, but I'm sorry in advance if I broke compatibility with other functions.

**Comment by fdschneider** [comment]
Great job. Benchmarking says your function is 150 times faster. Also the patch IDs are sequential while mine where quite random. 

Only difference between your and my function is the label 0 and NA for no patch, respectively. This does not matter much, if the downstream functions handle it correctly. But, maybe NA output of the `label()` function would be an easier way to achieve compatibility.
 
I added it to the label.R in 5c086a1

**Comment by fdschneider** [comment]
I just realized that your functions `psv()` and `psd()` are incompatible with this. Anyway they are redundant with `patchsizes()` and `indicator_cumpsd()` from ec1e55faa2c754cf842f80013d8a6b524acf1112.  (we were developing in parallel!)

**Comment by alexgenin** [comment]
Got it. 

Also, please do not forget to update tests when changing function behavior (fixed in 034b71c95124d57612dc2818447864d02ed6172b). I also moved the NA affectation in the cpp function so that the R function is just a wrapper (fixed in 82b92bd6af7e1aa37f4bf3d96a2bdcb048776ae4).
